applications/serial_lte_modem/src/slm_at_host.c

```diff
@@ -317,6 +333,81 @@ static int handle_at_slmuart(const char *at_cmd, uint32_t *baudrate)
 	return ret;
 }
 
+void silence_timer_handler(struct k_timer *timer)
+{
+	ARG_UNUSED(timer);
+
+	/* subordinate silence time-out */
+	exit_datamode();
+	if (slm_tcp_get_datamode()) {
+		slm_tcp_set_datamode_off();
+	}
+	if (slm_udp_get_datamode()) {
+		slm_udp_set_datamode_off();
+	}
+	datamode_off_pending = false;
+}
+
+K_TIMER_DEFINE(silence_timer, silence_timer_handler, NULL);
+
+static void raw_send(struct k_work *work)
+{
+	const char *terminator = CONFIG_SLM_DATAMODE_TERMINATOR;
+	int len = strlen(terminator);
+	int64_t silence_limit = CONFIG_SLM_DATAMODE_SILENCE * MSEC_PER_SEC;
+	int64_t silence_delta = k_uptime_delta(&silence_start);
+	int err;
+
+	ARG_UNUSED(work);
+
+	LOG_HEXDUMP_DBG(at_buf, at_buf_len, "RX");
+
+	/* check leading silence */
+	if (silence_delta < silence_limit) {
+		if (datamode_off_pending) {
+			uint8_t buf[UART_RX_LEN + len + 1];
+			/* cancel pending datamode off */
+			k_timer_stop(&silence_timer);
+			datamode_off_pending = false;
+			/* insert termiantor string */
+			strcpy(buf, terminator);
+			strcat(buf, at_buf);
```

Comment:
Isn't `at_buf` holding raw data? `strcat` would not work in this case.

Anyway, wouldn't it be simpler to `memmove()` the data to make space in the beginning for the terminator and then `memcpy()` the terminator string to the beginning? No need for an extra buffer on the stack then.

Reply:
_memmove()_ is what I'm looking for, and not found until you commented here :D